src/platform/crypt_openssl.c

```diff
@@ -463,50 +601,24 @@ CxPlatHashCreate(
     Hash->SaltLength = SaltLength;
     CxPlatCopyMemory(Hash->Salt, Salt, SaltLength);
 
-    Hash->Mac = EVP_MAC_fetch(NULL, "HMAC", NULL);
-    if (Hash->Mac == NULL) {
-        QuicTraceEvent(
-            LibraryError,
-            "[ lib] ERROR, %s.",
-            "EVP_MAC_fetch failed");
-        Status = QUIC_STATUS_TLS_ERROR;
-        goto Exit;
-    }
-
-    Hash->Ctx = EVP_MAC_CTX_new(Hash->Mac);
-    if (Hash->Ctx == NULL) {
-        QuicTraceEvent(
-            LibraryError,
-            "[ lib] ERROR, %s.",
-            "EVP_MAC_CTX_new failed");
-        Status = QUIC_STATUS_OUT_OF_MEMORY;
-        goto Exit;
-    }
-
     switch (HashType) {
     case CXPLAT_HASH_SHA256:
-        HashString = "sha256";
+        hctx = CXPLAT_HMAC_SHA256_CTX_HANDLE;
         break;
-    case CXPLAT_HASH_SHA384:
-        HashString = "sha384";
+     case CXPLAT_HASH_SHA384:
+        hctx = CXPLAT_HMAC_SHA384_CTX_HANDLE;
         break;
-    case CXPLAT_HASH_SHA512:
-        HashString = "sha512";
+     case CXPLAT_HASH_SHA512:
+        hctx = CXPLAT_HMAC_SHA512_CTX_HANDLE;
         break;
     default:
         Status = QUIC_STATUS_NOT_SUPPORTED;
         goto Exit;
     }
 
-    AlgParam[0] = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_DIGEST, (char*)HashString, 0);
-    AlgParam[1] = OSSL_PARAM_construct_end();
-
-    if (!EVP_MAC_CTX_set_params(Hash->Ctx, AlgParam)) {
-        QuicTraceEvent(
-            LibraryError,
-            "[ lib] ERROR, %s.",
-            "EVP_MAC_CTX_set_params failed");
-        Status = QUIC_STATUS_TLS_ERROR;
+    Hash->Ctx = EVP_MAC_CTX_dup(hctx);
+    if (Hash->Ctx == NULL) {
+        Status = QUIC_STATUS_OUT_OF_MEMORY;
```

Comment:
For allocation failures, please log the following type of event:
```
            QuicTraceEvent(
                AllocFailure,
                "Allocation of '%s' failed. (%llu bytes)",
                "EVP_MAC_CTX_dup",
                0); // If you don't know the size, just put zero
```
I'd appreciate it if you could add these to all initial points of allocation failure.

Reply:
Sure.